<div style="background: #000;
            color: #FFF;
            margin: 0px;
            padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1 >Week 18 Class 2 01/26</h1>
</div>

## Objectives for this week:
* Data Structuring 
* Data Cleanup

## Todays Agenda
* Continuing our code along
* Lab

<div style="background: #000;
            color: #FFF;
            margin: 0px;
            padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1 >Homework Problem</h1>
</div>

## Exercise 3

Given a stream of integers and a window size, calculate the moving average of all integers in the sliding window.

Implement the `MovingAverage` class:
* `MovingAverage(size)` Initializes the object with the size of the window size.
* `next(val)` Returns the moving average of the last size values of the stream.

**Example 1:** 
movingAverage = MovingAverage(3);  
movingAverage.next(1); // return 1.0 = 1 / 1  
movingAverage.next(10); // return 5.5 = (1 + 10) / 2  
movingAverage.next(3); // return 4.66667 = (1 + 10 + 3) / 3  
movingAverage.next(5); // return 6.0 = (10 + 3 + 5) / 3  
 

In [ ]:
import collections 

class MovingAverage(object):
    def __init__(self, size):
        """
        Initialize your data structure here.
        :type size: int
        """
        self.max_size = size
        self.window = []
        

    def next(self, val):
        """
        :type val: int
        :rtype: float
        """
        if len(self.window) == self.max_size:
            self.window.pop(0)
        self.window.append(val)
        return sum(self.window)/len(self.window)  

moving_average = MovingAverage(3)
print(moving_average.next(1))# == 1.0
print(moving_average.next(10))# == 5.5
print(moving_average.next(3))# == 4.666666666666667
print(moving_average.next(5))# == 6.0

try:
    moving_average = MovingAverage(3)
    assert(moving_average.next(1) == 1.0)
    assert(moving_average.next(10) == 5.5)
    assert(moving_average.next(3) == 4.666666666666667)
    assert(moving_average.next(5) == 6.0)
    moving_average = MovingAverage(5)
    assert(moving_average.next(4) == 4.0)
    assert(moving_average.next(1) == 2.5)
    assert(moving_average.next(10) == 5.0)
    assert(moving_average.next(5) == 5.0)
    print("solution is correct")
except AssertionError:
    print("solution is incorrect")

In [ ]:
from collections import deque

class MovingAverage(object):
    def __init__(self, size):
        """
        Initialize your data structure here.
        :type size: int
        """
        self.window = deque(maxlen=size)
        

    def next(self, val):
        """
        :type val: int
        :rtype: float
        """
        self.window.append(val)
        return sum(self.window)/len(self.window)  

try:
    moving_average = MovingAverage(3)
    assert(moving_average.next(1) == 1.0)
    assert(moving_average.next(10) == 5.5)
    assert(moving_average.next(3) == 4.666666666666667)
    assert(moving_average.next(5) == 6.0)
    moving_average = MovingAverage(5)
    assert(moving_average.next(4) == 4.0)
    assert(moving_average.next(1) == 2.5)
    assert(moving_average.next(10) == 5.0)
    assert(moving_average.next(5) == 5.0)
    print("solution is correct")
except AssertionError:
    print("solution is incorrect")


**[more on deque](https://docs.python.org/3/library/collections.html#collections.deque)** 

<div style="background: #000;
            color: #FFF;
            margin: 0px;
            padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1>Books</h1>
</div>

Start here:
[Mathematics for Computer Science](https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-042j-mathematics-for-computer-science-fall-2010/readings/MIT6_042JF10_notes.pdf)  
[Introduction to Probability](http://www.dartmouth.edu/~chance/teaching_aids/books_articles/probability_book/pdf.html)  
[Statistics](https://upload.wikimedia.org/wikipedia/commons/8/82/Statistics.pdf)  
[Think Stats (using python)](https://greenteapress.com/thinkstats2/thinkstats2.pdf)  
[Think Bayes](https://www.greenteapress.com/thinkbayes/thinkbayes.pdf)  
[Probabilistic Programming and Bayesian Methods for Hackers](https://nbviewer.jupyter.org/github/CamDavidsonPilon/Probabilistic-Programming-and-Bayesian-Methods-for-Hackers/blob/master/Chapter1_Introduction/Ch1_Introduction_PyMC3.ipynb)  

<div style="background: #000;
            color: #FFF;
            margin: 0px;
            padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1>Code Along</h1>
    <h3>Cleaning up dirty data</h3>
</div>

Some new things we'll cover in this code along:

[pandas.DataFrame.apply](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html) - applies a function to rows of a col  
[pandas.DataFrame.drop](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) - drops rows or columns  
[pandas.DataFrame.drop_duplicates](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) - drops duplicates  
[pandas.DataFrame.duplicated](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) - returns True if duplicate  
[pandas.DataFrame.fillna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html) - fills NaN values  
[pandas.merge](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) - merges two dfs  

Importing the libraries we'll be using

In [2]:
import numpy as np 
import pandas as pd 
import datetime

Let's open the `ted_talks.csv` file and generate a dataframe. Then we'll check the top of the `df` to see what we have.

In [3]:
df = pd.read_csv("../../datasets/ted_talks.csv")
df.head()

ID  comments                                        description  duration  \
0   0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1   1      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
2   2       265  With the same humor and humanity he exuded in ...       977   
3   3       124  New York Times columnist David Pogue takes aim...      1286   
4   4       200  In an emotionally charged talk, MacArthur-winn...      1116   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60       Emma Rob   
1  TED2006  1140825600         60       Emma Rob   
2  TED2006  1140825600         43            NaN   
3  TED2006  1140739200         26            NaN   
4  TED2006  1140912000         35  Majora Carter   

                                        name  num_speaker  published_date  \
0  Ken Robinson: Do schools kill creativity?            1      1151367060   
1  Ken Robinson: Do schools kill creativity?            1      1151367060   
2       Al Gore: Averting the climate crisis            1      1151367060   
3              David Pogue: Simplicity sells            1      1151367060   
4         Majora Carter: Greening the ghetto           -1      1151367060   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
2  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
3  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
4  [{'id': 3, 'name': 'Courageous', 'count': 760}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
3  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   

                   speaker_occupation  \
0                     Author/educator   
1                     Author/educator   
2                    Climate advocate   
3                Technology columnist   
4  Activist for environmental justice   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['children', 'creativity', 'culture', 'dance',...   
2  ['alternative energy', 'cars', 'climate change...   
3  ['computers', 'entertainment', 'interface desi...   
4  ['MacArthur grant', 'activism', 'business', 'c...   

                         title  \
0  Do schools kill creativity?   
1  Do schools kill creativity?   
2  Averting the climate crisis   
3             Simplicity sells   
4          Greening the ghetto   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
2  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
3  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
4  https://www.ted.com/talks/majora_carter_s_tale...   1697550

Checking the shape of the `df`

In [6]:
df.shape

(2555, 18)

Checking info to see other information (like how much RAM our dataset uses and how many non-null entries we have)

In [ ]:
df.info()

We see we have 2555 rows but 2531 non-null rows for 'main_speaker'. So let's check that out.

In [ ]:
df[df['main_speaker'].isna()][["description", "main_speaker","name"]]

Before we fix the "main_speaker" column, let's shrink our dataset by removing duplicates. 
First we need to see how many duplicates we have:

In [ ]:
df['url'].duplicated().sort_values(ascending=False)[:6]

In [ ]:
df[(df['url'].duplicated() == False)]

In [ ]:
df[df['url'].duplicated()]

Then we'll drop the duplicates using the `df.drop_duplicates` method. 

Remember, the keyword `inplace` means it will mutate the original `df`.

In [ ]:
help(df.drop_duplicates)

In [ ]:
df.drop_duplicates('url', inplace=True)

Next we'll drop the `ID` column as it's unnecessary

In [ ]:
df.drop('ID', axis=1, inplace=True)

If we look at `published_date` and `film_date`, they're formatted as unix timestamps. We might need a human-readable format.

[more on unix timestamps](https://en.wikipedia.org/wiki/Unix_time)

In [ ]:
df[["published_date", "film_date"]]

Let's see how we can turn a unix timestamp into a human-readable format

In [ ]:
ts = 1151367060

s = datetime.datetime.fromtimestamp(ts)

s.strftime("%m/%d/%Y, %H:%M:%S")

Putting that together, we can create a function that does this for any timestamp

In [ ]:
def transform_date(timestamp):
    _date = datetime.datetime.fromtimestamp(timestamp)
    return _date.strftime("%m/%d/%Y, %H:%M:%S")

And then we can test it with a random timestamp

In [ ]:
transform_date(1151367060)

We can use the `df.apply` method to apply that function to each row of a specific column.

In [ ]:
df['film_date'].apply(transform_date)

Seeing that it works, we can now use that `transform_date` method along with `apply` to create two new columns of human-readable timestamps

In [ ]:
df['film_date_hr'] = df['film_date'].apply(transform_date)
df['published_date_hr'] = df['published_date'].apply(transform_date)

If we wanted to change the ordering of the columns, we can use `df.columns.tolist()` to list our columns and then `df.reindex()` to reindex the columns by passing in a list of columns in the order we want

In [ ]:
df.columns.tolist()

In [ ]:
df = df.reindex(['comments',
 'description',
 'duration',
 'event',
 'film_date',
 'film_date_hr',
 'languages',
 'main_speaker',
 'name',
 'num_speaker',
 'published_date',
 'published_date_hr',
 'ratings',
 'related_talks',
 'speaker_occupation',
 'tags',
 'title',
 'url',
 'views',
],axis=1)

We can see that works by checking the head of the `df`

In [ ]:
df.head()

Next, we'll fix the durations, which are in seconds, by converting them to mins and rounding the result to 2 places after the decimal

In [ ]:
df['duration'] = round( df['duration'] / 60, 2)

We'll fix incorrect values for `languages` and `num_speakers` by replacing erroneous values with `1` since we know there's at least one language translated (english) and there's at least one speaker.

In [ ]:
df[df['languages'] == 0]

In [ ]:
df['languages'].replace(0, 1, inplace=True)
df

In [ ]:
df[df['languages'] == 0]

In [ ]:
df[(df['num_speaker'] <= 0)]

In [ ]:
df["num_speaker"].replace(0, 1, inplace=True)
df["num_speaker"].replace(-1, 1, inplace=True)

Finally, we're able to get back to the first issue of `main_speaker` having blanks. We can use `df['main_speaker'].isna()` to check the rows of `main_speaker` that are empty

In [ ]:
df['main_speaker'].isna()

If we look at the `description`, `main_speaker` and `name` columns for all of the blank rows, we see some useful information

In [ ]:
df[df['main_speaker'].isna()][["description", "main_speaker","name"]]

Without that useful information, we can always just replace empty values with a default value like so:

In [ ]:
df.fillna("unknown", inplace=True)

In [ ]:
df

In [ ]:
df[df['main_speaker'] == 'unknown'].shape

But here, `name` also contains the speaker's name. So we can rip it and then adjust `main_speaker` and `name` accordingly

In [ ]:
df['name'].head()

In [ ]:
df['main_speaker'] = df['name'].str.split(':',1).str[0]

In [ ]:
df['name'] = df['name'].str.split(':',1).str[1]

In [ ]:
df

<div style="background: #000;
            color: #FFF;
            margin: 0px;
            padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1>Lab</h1>
</div>

For today's lab, we'll be using the Marvel comic book characters data set available from [538](https://github.com/fivethirtyeight/data/tree/master/comic-characters)

The data comes from [Marvel Wikia](http://marvel.wikia.com/Main_Page) and [DC Wikia](http://dc.wikia.com/wiki/Main_Page). Characters were scraped on August 24. Appearance counts were scraped on September 2. The month and year of the first issue each character appeared in was pulled on October 6.

The data is split into two files, for DC and Marvel, respectively: `dc-wikia-data.csv` and `marvel-wikia-data.csv`. Each file has the following variables:

Variable | Definition
---|---------
`page_id` | The unique identifier for that characters page within the wikia
`name` | The name of the character
`urlslug` | The unique url within the wikia that takes you to the character
`ID` | The identity status of the character (Secret Identity, Public identity, [on marvel only: No Dual Identity])
`ALIGN` | If the character is Good, Bad or Neutral
`EYE` | Eye color of the character
`HAIR` | Hair color of the character
`SEX` | Sex of the character (e.g. Male, Female, etc.)
`GSM` | If the character is a gender or sexual minority (e.g. Homosexual characters, bisexual characters)
`ALIVE` | If the character is alive or deceased
`APPEARANCES` | The number of appareances of the character in comic books (as of Sep. 2, 2014. Number will become increasingly out of date as time goes on.)
`FIRST APPEARANCE` | The month and year of the character's first appearance in a comic book, if available
`YEAR` | The year of the character's first appearance in a comic book, if available


Importing the libraries we'll be using

In [2]:
import numpy as np 
import pandas as pd 
import datetime

Let's open the `marvel-wikia-data.csv` file and generate a dataframe. Then we'll check the top of the `df` to see what we have.

In [3]:
df = pd.read_csv("../../datasets/marvel-wikia-data.csv")
df.head()

page_id                                 name  \
0     1678            Spider-Man (Peter Parker)   
1     7139      Captain America (Steven Rogers)   
2    64786  Wolverine (James \"Logan\" Howlett)   
3     1868    Iron Man (Anthony \"Tony\" Stark)   
4     2460                  Thor (Thor Odinson)   

                                   urlslug                ID  \
0              \/Spider-Man_(Peter_Parker)   Secret Identity   
1        \/Captain_America_(Steven_Rogers)   Public Identity   
2  \/Wolverine_(James_%22Logan%22_Howlett)   Public Identity   
3    \/Iron_Man_(Anthony_%22Tony%22_Stark)   Public Identity   
4                    \/Thor_(Thor_Odinson)  No Dual Identity   

                ALIGN         EYE        HAIR              SEX  GSM  \
0     Good Characters  Hazel Eyes  Brown Hair  Male Characters  NaN   
1     Good Characters   Blue Eyes  White Hair  Male Characters  NaN   
2  Neutral Characters   Blue Eyes  Black Hair  Male Characters  NaN   
3     Good Characters   Blue Eyes  Black Hair  Male Characters  NaN   
4     Good Characters   Blue Eyes  Blond Hair  Male Characters  NaN   

               ALIVE  APPEARANCES FIRST APPEARANCE    Year  
0  Living Characters       4043.0           Aug-62  1962.0  
1  Living Characters       3360.0           Mar-41  1941.0  
2  Living Characters       3061.0           Oct-74  1974.0  
3  Living Characters       2961.0           Mar-63  1963.0  
4  Living Characters       2258.0           Nov-50  1950.0

As you can see we have a couple of issues:
* The column `names` has, in some cases, both the hero name and their secret identity. In other cases, it has the universe the hero is from (usually Earth-616). We'd like for `name` to be the superhero. If a secret identity name is available, we'd like for that to be in it's own column. We'd also like a column `origin` for the superhero's place of origin (usually Earth-616).
* The `urlslug` column is incorrect. We'd like for that to be a complete link. For example, spiderman's wikia link is https://marvel.fandom.com/wiki/Peter_Parker_(Earth-616) but if you use the `Spider-Man_(Peter_Parker)` from the `urlslug` (https://marvel.fandom.com/wiki/Spider-Man_(Peter_Parker), the page redirects to the correct page. You can choose to fix it various different ways but it should be a working link. Use "https://marvel.fandom.com/wiki/" as the beginning of your url.
* For the column `align`, we'd like for the "Characters" part to be removed. We'd like for that column to be categorical (Good, Neutral, Evil or blank).
* Same for the columns `EYE`, `HAIR`, `SEX` and `ID`. We want to drop the extraneous data and have simple categories (e.g. for `ID` we want categories ["Known","Public", "Secret", "None"] and for `SEX` we just want ["Male", "Female","Genderfluid","Agender"]. There's a way shown below that shows the various possible categories.
* Some data is in the wrong format. For example, `Year` and `APPEARANCES` are floats but they really should be ints (as you cannot have a non-int number of appearances). 
* Some data doesn't match up (like the `FIRST APPEARANCE` and `Year` columns might have not be the same year)
* Some columns are unnecessary. Some are missing data. Some rows are duplicates. Fix those as necessary.
* Some column names are written as uppercase, some are lowercase and some are mixed case. Normalize them so all the column names are in the same format.

Fix as many of these issues as you're able to. At the minimum, complete 4 of the above bulletpoints.

In [ ]:
def see_possible_categories(col):
    return df.groupby(col).sum()
    
see_possible_categories('name')